balancer/rls/internal/config.go

```diff
@@ -109,37 +105,49 @@ type lbConfigJSON struct {
 	ChildPolicyConfigTargetFieldName string
 }
 
-// ParseConfig parses and validates the JSON representation of the service
-// config and returns the loadBalancingConfig to be used by the RLS LB policy.
-//
-// Helps implement the balancer.ConfigParser interface.
-// * childPolicy field:
-//  - must find a valid child policy with a valid config (the child policy must
-//    be able to parse the provided config successfully when we pass it a dummy
-//    target name in the target_field provided by the
-//    childPolicyConfigTargetFieldName field)
-// * childPolicyConfigTargetFieldName field:
+// When parsing a config update, the following validations are performed:
+// - routeLookupConfig:
+//   - grpc_keybuilders field:
+//     - must have at least one entry
+//     - must not have two entries with the same `Name`
+//     - within each entry:
+//       - must have at least one `Name`
+//       - must not have a `Name` with the `service` field unset or empty
+//       - within each `headers` entry:
+//         - must not have `required_match` set
+//         - must not have `key` unset or empty
+//       - across all `headers`, `constant_keys` and `extra_keys` fields:
+//         - must not have the same `key` specified twice
+//         - no `key` must be the empty string
+//   - `lookup_service` field must be set and and must parse as a target URI
+//   - if `max_age` > 5m, it should be set to 5 minutes
+//   - if `stale_age` > `max_age`, ignore it
+//   - if `stale_age` is set, then `max_age` must also be set
+//   - ignore `valid_targets` field
+//   - `cache_size_bytes` field must have a value greater than 0, and if its
+//      value is greater than 5M, we cap it at 5M
+// - childPolicy:
+//   - must find a valid child policy with a valid config
+// - childPolicyConfigTargetFieldName:
 //   - must be set and non-empty
-func (*rlsBB) ParseConfig(c json.RawMessage) (serviceconfig.LoadBalancingConfig, error) {
+func parseConfig(c json.RawMessage) (serviceconfig.LoadBalancingConfig, error) {
```

Comment:
Why this change?  It seems to complicate this control flow unnecessarily by adding in an extra call into the stack.

Reply:
The builder does not have any state that is used in this function, hence I thought there is no need for this function to have a receiver. Also, we do have usages like this currently in the codebase (although that is not a good enough reason).

Also, I find it easier to navigate through the code when a single type's methods are not spread across multiple files. Splitting the config parsing logic this way helps put all the config parsing logic in one file, and the builder implementation is just boiler plate.

Also, wouldn't the compiler be able to optimize this and inline the function?

Reply:
> The builder does not have any state that is used in this function, hence I thought there is no need for this function to have a receiver.

This is when you omit the variable for the receiver, though.

> Also, I find it easier to navigate through the code when a single type's methods are not spread across multiple files. Splitting the config parsing logic this way helps put all the config parsing logic in one file, and the builder implementation is just boiler plate.

Adding more code and more functions to make it easier to navigate?  I think naming the file `config.go` is probably good enough.  Or move the implementation of `ParseConfig` into `builder.go` as the other option?

> Also, wouldn't the compiler be able to optimize this and inline the function?

I'm not really concerned about performance.  We don't parse configs very often.  It just seems unnecessary to have this function here do ~nothing and call another function in the same package.  If you want to signal that the builder implements it, add a `var _ balancer.ConfigParser = rlsBB{}` line near the `rlsBB` struct definition?

Reply:
Done.